### ISSUE TITLE
Exception executing HQL query with uncorrelated left joins in subselect

### ISSUE THREAD

**Issue by csharper2010** [nhibernate/nhibernate-core#3334]
### The base situation

We have a data model like this one
```
Entity ---0..* ChildEntity
       ---0..1 OtherEntity

ChildEntity ---1 GrandChildEntity
```

The associations to OtherEntity and GrandChildEntity are mapped as ```many-to-one```.

When executing the HQL query ...
```sql
SELECT ROOT 
FROM Entity AS ROOT 
WHERE
EXISTS 
    (FROM ELEMENTS(ROOT.Children) AS child
        LEFT JOIN child.Child AS grandChild
        LEFT JOIN ROOT.OtherEntity AS otherEntity
    WHERE
        grandChild.Name like 'G%'
        OR otherEntity.Name like 'G%'
    )
```

### The problem

... we get the following sql generated:
```sql
select entity0_.Id as id1_0_3_, entity0_.Name as name2_0_3_, entity0_.OtherEntity as otherentity3_0_3_
from Entity entity0_ 
where exists (
    select children1_.Id, grandchild2_.Id, otherentit3_.Id
    from ChildEntity children1_
        left outer join GrandChildEntity grandchild2_ on children1_.Child=grandchild2_.Id, 
        left outer join OtherEntity otherentit3_ on entity0_.OtherEntity=otherentit3_.Id 
    where entity0_.Id=children1_.Parent and (grandchild2_.Name like 'G%' or otherentit3_.Name like 'G%'))
```

This leads to an exception because the comma ```,``` before the last ```left outer join``` is illegal.

### Side note on why we don't use direct dot-navigations instead of explicit joins

Using implied joins on the :1-navigations like in the following query does not work because then, the joins are created as ```inner join``` or theta join so the condition is not evaluated correctly if there is not OtherEntity associated with Entity:
```sql
SELECT ROOT 
FROM Entity AS ROOT 
WHERE
    EXISTS 
        (FROM ELEMENTS(ROOT.Children) AS child
        WHERE
        child.Child.Name like 'G%'
        OR ROOT.OtherEntity.Name like 'A%'
)
```

I might start a separate discussion on Nullability and ```many-to-one``` because of that.

**Comment by bahusoid** [comment]
```SQL
 LEFT JOIN ROOT.OtherEntity AS otherEntity
```
I would say it's invalid hql. What this join does in subquery? It makes no sense to me.
Why is it not part of main query?

```sql
SELECT ROOT 
FROM Entity AS ROOT 
 LEFT JOIN ROOT.OtherEntity AS otherEntity -- MOVED HERE
WHERE
EXISTS 
    (FROM ELEMENTS(ROOT.Children) AS child
        LEFT JOIN child.Child AS grandChild       
    WHERE
        grandChild.Name like 'G%'
        OR otherEntity.Name like 'G%'
    )
```

**Comment by csharper2010** [comment]
> I would say it's invalid hql. What does this join do in subquery? It makes no sense to me.
> Why is it not part of main query?

HQL documentation does not mention any restrictions on such joins and the same way you can place outer joins in SQL wherever you want, it could be handled in HQL.

That query worked fine in NHibernate 1.2.1. and it is not hand-written HQL but generated from our application.

**Comment by bahusoid** [comment]
It's another case of  https://github.com/nhibernate/nhibernate-core/issues/1228 

It's actually mentioned in original issue discussion that provided hql looks wrong. I also already mentioned it: https://github.com/nhibernate/nhibernate-core/pull/2146#issuecomment-1013669170

**Comment by csharper2010** [comment]
Yep, I also have a feeling that my fix won't handle all the posssible cases...
The tree generation is very complex and I don't have the knowledge about corner cases that may exist so it's hard to estimate all the consequences of a change.

It would be ideal if the code in #2146 would not be required, it was a workaround already then but it was driven by the comment that the from clause processor will take care of adding it to the right place.

I would be willing to do some further steps but it's really not an easy task because of those complex interdependencies.

**Comment by csharper2010** [comment]
Now I had some time to work on a cleaner fix: #3338

@bahusoid: the query you gave in your comment now also works because I do the translation based on the explicit case where necessary.
